runtime/interface.go

```diff
@@ -41,55 +43,130 @@ type Block struct {
+// AccountStorage stores and retrives account key/values
+type AccountStorage interface {
+	// Value gets a value for the given key in the storage, owned by the given account.
+	Value(key StorageKey, caller Location) (value StorageValue, err error)
+	// SetValue sets a value for the given key in the storage, owned by the given account.
+	SetValue(key StorageKey, value StorageValue, caller Location) (err error)
+	// ValueExists returns true if the given key exists in the storage, owned by the given account.
+	ValueExists(key StorageKey, caller Location) (exists bool, err error)
+	// StoredKeys returns list of keys and their sizes owned by this account
+	StoredKeys(address Address, caller Location) (iter StorageKeyIterator, err error)
+	// StorageUsed gets storage used in bytes by the address at the moment of the function call.
+	StorageUsed(address Address, caller Location) (value uint64, err error)
+	// Note: StorageCapacity has been moved to injected methods (similar to get balance)
+}
+
+// AccountKeys manages account keys
+type AccountKeys interface {
+	// AddAccountKey appends a key to an account.
+	AddAccountKey(address Address, publicKey []byte, caller Location) error
+	// RemoveAccountKey removes a key from an account by index.
+	RevokeAccountKey(address Address, index int, caller Location) (publicKey []byte, err error)
+	// AccountPublicKey returns the account key for the given index
+	AccountPublicKey(address Address, index int, caller Location) (publicKey []byte, err error)
+}
+
+type LocationResolver interface {
 	// GetCode returns the code at a given location
 	GetCode(location Location) ([]byte, error)
+	// ResolveLocation resolves an import location.
+	ResolveLocation(identifiers []Identifier, location Location) ([]ResolvedLocation, error)
+}
+
+// Results are responsible to capture artifacts generated
+// when running the runnable
+//
+// Results won't be directly callable by users
+// and will be used by the runtime to capture outputs
+type Results interface {
+	// AppendLog appends a log to the log collection
+	AppendLog(string) error
+	// Logs returns all the logs
+	Logs() ([]string, error)
+	// returns log i of the log collection
+	LogAt(i uint) (string, error)
+	// returns number of logs
+	LogCount() uint
+
+	// AppendEvent appends an event to the event collection
+	AppendEvent(cadence.Event) error
+	// Events returns all the events
+	Events() ([]cadence.Event, error)
+	// returns event i of the event collection
+	EventAt(i uint) (cadence.Event, error)
+	// returns number of events
+	EventCount() uint
+
+	// AppendError appends a non-fatal error
+	AppendError(error) error
+	// return all the errors
+	Errors() multierror.Error
+	// returns event i of the event collection, the first error is the actual runtime error, the second error
+	// returns if there is any error while fetching the error i
+	ErrorAt(i uint) (Error, error)
+	// returns number of errors
+	ErrorCount() uint
```

Comment:
What is calling this append error function? Do you have an example use-case? 
Also, why is it a collection, wouldn't the program abort on the first error, as it does right now?

Reply:
yeah, just keeping some room for the future in case we want to return all the errors. but if it will always be one error I can change it.